### ISSUE TITLE
Add 'split' support for ReadOnlySpan<char> similar to string

### ISSUE THREAD

**Issue by ahsonkhan** [dotnet/runtime#934]
**Latest Update:**

**New proposal to re-review:** Enumerator returns `Range` instead of `ReadOnlySpan<T>`
```C#
partial class MemoryExtensions
{
    public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> span, T separator,
        StringSplitOptions options = StringSplitOptions.None) where T : IEquatable<T> {}

    public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
    {
        public SpanSplitEnumerator<T> GetEnumerator() { return this;  }
        public bool MoveNext();
        public Range Current { get; }
    }
}
```

**Previously approved API Proposal**
```C#
public static SpanSplitEnumerator<T> Split(this ReadOnlySpan<T> span, T seperator,
    StringSplitOptions options = StringSplitOptions.None) where T : IEquatable<T> {}

public ref struct SpanSplitEnumerator<T> where T : IEquatable<T>
{
    public SpanSplitEnumerator GetEnumerator() { return this;  }
    public bool MoveNext();
    public ReadOnlySpan<T> Current { get; }
}
```

Split off from https://github.com/dotnet/corefx/issues/21395#issuecomment-359342832

_From @Joe4evr on January 21, 2018 23:16_

> Can I throw in another suggestion? I'd really like to see some ability to split a `ReadOnlySpan<char>`. Obviously, you can't return a collection of `Span`s directly, but isn't that what `Memory<T>` is for?
```cs
// equivalent to the overloads of 'String.Split()'
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, char[] seperator);
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, char[] seperator, int count);
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, char[] seperator, int count, StringSplitOptions options);
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, char[] seperator, StringSplitOptions options);
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, string[] seperator, int count, StringSplitOptions options);
public static IReadOnlyList<ReadOnlyMemory<char>> Split(this ReadOnlySpan<char> span, string[] seperator, StringSplitOptions options);
```
> The reason for choosing `IReadOnlyList<T>`  is to make the resulting collection indexable, just like `string[]`. It would be nice if the implementation is `ImmutableArray<T>`, but I'm not sure if that's a concern for distribution and such.
---

_From @ahsonkhan on January 22, 2018 01:36_
> @Joe4evr, out of curiosity, do you have a scenario atm where these APIs would be useful? If so, can you please show the code sample?

> I would replace the char[] overloads with ReadOnlySpan\<char\>. However, I am not sure about adding the split APIs, in general, given they have to allocate. Is there a way to avoid allocating? Also, given these are string-like APIs for span, it is strange to have an overload that takes a string[]. Maybe all these would fit better on ReadOnlyMemory instead, especially given the return type.
---

_From @Joe4evr on January 22, 2018 08:35_
> My scenario is taking a relatively big string of user input and then parsing that to populate a more complex object. So rather than take the whole string at once, I'd like to parse it in pieces at a time. It'd be pretty nice if this can be facilitated by the `Span`/`Memory<T>` APIs so that this code won't have to allocate an extra 30-40 tiny strings whenever it runs.

> Admittedly, I only started on this particular case earlier today, mostly to experiment and find out how much I could get out of the Span APIs at this time.

> Maybe it was a bit naive of me to expect a collection like I did, but I'd at least like to see some API to deal with this scenario a little easier, because I'll probably not be the only one looking to split a span up into smaller chunks like this.
---

_From @stephentoub on January 22, 2018 08:41_
> Splitting support would be good, but I don't think it would look like the proposed methods; as @ahsonkhan points out, that would result in a lot of allocation (including needing to copy the whole input string to the heap, since you can't store the span into a returned interface implementation).

> I would instead expect a design more like an iterator implemented as a ref struct, e.g.
```C#
public ref struct CharSpanSplitter
{
    public CharSpanSplitter(ReadOnlySpan<char> value, char separator, StringSplitOptions options);
    public bool TryMoveNext(out ReadOnlySpan<char> result);
}
```

cc @KrzysztofCwalina, @stephentoub, @Joe4evr

**Comment by danmosemsft** [comment]
@bbartels , dotnet/runtime is open for business :)

**Comment by bbartels** [comment]
@danmosemsft I really should have commented here, but already submitted a PR a couple weeks ago! https://github.com/dotnet/runtime/pull/295

**Comment by jonathanmarston** [comment]
Ran into a good use case for this just today. Would love to see this finished up. 

One minor nitpick though: doesn't having a single type acting as both enumerable and enumerator create a less-than-obvious inconsistency with the rest of .NET?

```csharp
var split = "a,b,c".Split(',');

foreach (var letter in split) 
{
    // Hit three times
} 

foreach (var letter in split) 
{
    // Hit three times 
}

var splitSpan = "a,b,c".AsSpan().Split(',');

foreach (var letter in splitSpan) 
{
    // Hit three times
} 

foreach (var letter in splitSpan) 
{
    // Never hit because splitSpan was already consumed by first foreach
}
```

**Comment by SeeminglyScience** [comment]
@jonathanmarston 

> One minor nitpick though: doesn't having a single type acting as both enumerable and enumerator create a less-than-obvious inconsistency with the rest of .NET?

The compiler creates a copy of the struct enumerator before enumerating ([SharpLab Example](https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetE+WMm2uxAbgIYBOiApogLyIB21A7ogMoAuFArgMacAoox4BbahTKcA9hUEAPMqIAOAG2oAKeAAYAlAG4ixHADNZ1MnwAWiDQEtGnRHeeMauo8awxjv5BABOewNPYgBfWFDcMwoLa1sHJxcHdyicbz9jEEDgwx8/CPycQuJPBERYk0QAZ25+Jy5eAWExCSlZBSU1Wgy/ZQo7ck5aRMQAfQcAE2p5PMy0xH7BqVpYsknpRlUAT1cnMb5pHkc5vwXyxvqW8UkZOUUVdXtHREPjzg8irwXiA6OXhhvE4/XATRjTeT0RAAWggp18hXOSAARtJpKpEABZaQkagAORmnA0n0yvUyODsVQ0YIhiAAfAw/u9dMgAOyIExkVTVajw8kgnA0mYAamFfMyIHZdV5CxKZy+xHKo1QPAosQBdPGUxm4twSOQABZEAAlag8oksuiaoWQhiw3U4fWXZoiG7te5ddSIADi1CErrad2J9E1nCsdmqfMKYSAA==)).

**Comment by jonathanmarston** [comment]
> The compiler creates a copy of the struct enumerator before enumerating
Ah yes. I see it now. Thanks!

**Comment by wstaelens** [comment]
is this already available? 🥺

**Comment by bbartels** [comment]
@wstaelens Still waiting for the PR (https://github.com/dotnet/runtime/pull/295) to be reviewed/approved, but feel free to copy the implementation into your codebase if you need it early.

**Comment by GrabYourPitchforks** [comment]
Reopening since we'll need to send this through review again.

**Comment by wstaelens** [comment]
any news on ROS.split(',') ?

**Comment by loop-evgeny** [comment]
What's the status of this? It looked like it was merged into .NET 5.0, but it's not there, nor mentioned in the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.memoryextensions?view=net-5.0).

**Comment by GrabYourPitchforks** [comment]
@loop-evgeny It was reverted before 5.0 shipped. There's currently no public implementation.

**Comment by wstaelens** [comment]
> @loop-evgeny It was reverted before 5.0 shipped. There's currently no public implementation.

😞 🐼

**Comment by loop-evgeny** [comment]
OK, thanks, but just to confirm: there was nothing wrong with the implementation per se, and it was just the API that you want to revise? So the code in that PR can be used if it works for me?

**Comment by wstaelens** [comment]
> OK, thanks, but just to confirm: there was nothing wrong with the implementation per se, and it was just the API that you want to revise? So the code in that PR can be used if it works for me?

when can we expect it ?

if it isn't there soon, which code can we already implement?

why was it pulled back?

**Comment by bbartels** [comment]
> OK, thanks, but just to confirm: there was nothing wrong with the implementation per se, and it was just the API that you want to revise? So I can use the code in that PR if it works for me?

The implementation works as long as you expect the implementation to only split on the space character (as string.Split() splits on all whitespace characters).
More information here:  https://github.com/dotnet/runtime/pull/295#issuecomment-642394483

Also, since the implementation uses ROS.IndexOf, which while vectorised, has some overhead and may result in slower performance than String.Split after this commit (especially when separators density is high): https://github.com/dotnet/runtime/pull/38001#issuecomment-740321986.

I was going to put another API proposal forward that may address the concern of string.Split being faster than SpanSplitEnumerator, in that it has a buffer of separator indices that once filled will return from MoveNext().

**Comment by wstaelens** [comment]
looking forward to it as I in converting code the ROS.split on ' ' and '\t' would be very very useful 👍

**Comment by davidfowl** [comment]
This is is the "Future" milestone, which means it isn't happening. I actually don't recall what the issues were with the original change.

**Comment by davidfowl** [comment]
Here it is https://github.com/dotnet/runtime/pull/37742#issuecomment-642736794. @GrabYourPitchforks was there another proposal to solve this?

**Comment by Symbai** [comment]
I prefer the name `Enumerate` more than `Split`. Because of the similarity of `String.Split` and `ReadOnlySpan<char>.EnumerateLines()`. On `Split` people would assume to retrieve an array. `Enumerate` is more clear when its just an enumerator and aligns better to the existing `EnumerateLines()`, see #68487

**Comment by bbartels** [comment]
### TLDR

- Build a separate types for splitting of `ROS<char>`and `ROS<TElement>`
- Performance could be improved with a different approach but will result in a worse API and therefore (IMO), not worth it (see discussion [below](#performance-discussion))
- As the perf-discussion is quite long, and not necessary for api review related discussion you can skip to [Api Review notes](#api-review) and [Api Proposal](#api-proposal)

## Performance Discussion

Hi everyone!

Apologies this took so long, after finishing university I was just incredibly busy with work and I found it difficult to devote time to this. But I spent the last couple months prototyping some solutions to the questions outlined [here](https://github.com/dotnet/runtime/issues/934#issuecomment-829250284).  
As the [original prototype](https://github.com/dotnet/runtime/pull/295) used `ROS.IndexOf` instead of a custom SIMD implementation to retrieve multiple separator indices in one call, there were concerns about performance of the type.
`string.Split` now uses such a [custom SIMD implementation](https://github.com/dotnet/runtime/pull/38001) that can retrieve multiple separator indices in one go, which was found to be faster than just using `ROS.IndexOf`.

The prototypes I built tried to address some of the peformance concerns of separator-dense strings, to see whether it may be possible to build a type that does perform better in such scenarios.

The main idea for improvements is to reduce the number of calls to `ROS.IndexOf` and implement an algorithm that is similar to the one outlined [here](https://github.com/dotnet/runtime/pull/38001), where multiple indices are accumulated using SIMD.
This runs into two issues, as a generic implementation is difficult if not impossible due to varying element sizes taking up varying amounts of space within a vector register. The second being that one loses the ability to do granular paging through the source sequence, as the buffer needs to be filled in one pass before being able to return elements to the caller. This can result in worse runtime performance if one intends to terminate the enumeration early.

While these are grave issues there are ways to work around them. One could have fast paths for specific element sizes, and have a buffered and unbuffered type to enable both scenarios.
To investigate what kind of improvements a buffered approach would yield I implemented a few protopytes. The improvements seen using a buffered prototype that utilises the same SIMD algorithm as used in `string.Split` showed some improvements
(15-25%) over the original prototype. Improvements were generally found in circumstances of high separator density, while (slight) regressions were found in the opposite case. It seems that the buffered approach has some overhead that is only negated when a separator dense string is encountered.

The second prototype was more of an experiment to see what is possible in terms of performance. The largest overhead found with the buffered approach (apart from calling MoveNext) is the additional abstraction of `Range` in `Current`. To alleviate the perf cost of this abstraction I changed `Current` to return a `ReadOnlySpan<int>` which contains a raw chunk of accumulated separators. While this type is superior to both the aforementioned buffered approach and the original `SpanSplitEnumerator` the API is not exactly great. Needing two loops to iterate over the results is not very user friendly. Additionally the relative performance improvement gets less significant the more work is being done using each separator index. 

Therefore, while both types would improve performance, one only does so when the source string is dense with separators, and the other has a very inconvenient API, I would advise against adding a buffered (or chunked) approach.

In case someone is interested in playing with the prototypes I've pushed them to [this](https://github.com/bbartels/SpanSplitEnumerator) repository.  
Just a warning they are completely untested so proceed with caution. The ChunkSplitEnumerator might actually be interesting to folk who need the absolute fastest way to enumerate seperator indices.


## Api Review

As @GrabYourPitchforks points out [here](https://github.com/dotnet/runtime/pull/37742#issuecomment-642736794), there are numerous concerns about the previously approved API. First, the feature-set of a `string` based `SpanSplitEnumerator` needs to be carefully considered.

Concerns about feature-set:

- Splitting on any whitespace char to align implementation closer to `string.Split`
    - Change implementation to respect `char.IsWhitespace`
- Splitting on single char
    - As per previous proposal
- Splitting on a multi-char substring
    - As per previous proposal

Additionally, features outlined below need to be discussed as well (although previous review consensus leaned towards delaying these features until a use case is apparent)

- Splitting on any of a set of multi-char substrings
- Splitting on any of a set of chars
- All of the above, but with a `StringSplitOptions`


What needs to be discussed in the API review: 

- Discuss the feature set that a string based `SpanSplitEnumerator` should exhibit.
- Discuss whether to have discrete implementations for a String based `SpanSplitEnumerator` and a generic one.
- Discuss whether the previously accepted generic `SpanSplitEnumerator<T>` API should be implemented to complement a String based one (assuming discrete implementations).
- Discuss as @Symbai mentioned to change the name of Split to Enumerate [see here](https://github.com/dotnet/runtime/issues/934#issuecomment-1108632073).

I understand that there is a _lot_ of context to consider, so if there is desire for me to be present during review I'd be happy to do so (assuming external contributors are allowed on call).

## Api Proposal

Proposal for discrete string based split enumerator
```csharp
public static class MemoryExtensions
{
    public static StringSplitEnumerator Split(this ReadOnlySpan<char> span)
        // Would use char.IsWhitespace as default separator (instead of string.Empty)
        => new StringSplitEnumerator(span);
//or    => new StringSplitEnumerator(span, string.Empty) depending on discussion

    public static StringSplitEnumerator Split(this ReadOnlySpan<char> span, char separator)
        => new StringSplitEnumerator(span, separator);

    public static StringSplitEnumerator Split(this ReadOnlySpan<char> span, string separator)
    // What should happen if null is passed, align with char.IsWhiteSpace, string.Empty or throw Exception?
        => new StringSplitEnumerator(span, separator ?? string.Empty);
}

public ref struct StringSplitEnumerator
{
    StringSplitEnumerator GetEnumerator();
    bool MoveNext();
    Range Current { get; }
}
```

Proposal for generic span split enumerator
```csharp
public static class MemoryExtensions
{
    public static SpanSplitEnumerator<T> Split(this ReadOnlySpan<T> span, T separator)
        => new SpanSplitEnumerator<T>(span, separator);

    public static SpanSplitEnumerator<T> Split(this ReadOnlySpan<T> span, ReadOnlySpan<T> separator)
        => new SpanSplitEnumerator<T>(span, separator);

    // Would an overload for split on `default` make sense?
    public static SpanSplitEnumerator<T> Split(this ReadOnlySpan<T> span)
        => new SpanSplitEnumerator<T>(span, default);
}

public ref struct SpanSplitEnumerator<T>
{
    SpanSplitEnumerator<T> GetEnumerator();
    bool MoveNext();
    Range Current { get; }
}
```

**Comment by bbartels** [comment]
With #76186 being approved now is there still interest in continuing work on this? It does serve a slightly different purpose. If yes this could be marked as ready for review @stephentoub

**Comment by stephentoub** [comment]
> is there still interest in continuing work on this?

Yup, thanks.

@GrabYourPitchforks, do you have any thoughts on @bbartels's proposal and open issues?

**Comment by GF-Huang** [comment]
This things seems good: https://www.meziantou.net/split-a-string-into-lines-without-allocation.htm

**Comment by bbartels** [comment]
Hi @GrabYourPitchforks,

just a gentle bump on this :)

**Comment by bbartels** [comment]
@stephentoub @GrabYourPitchforks  Any chance we could get some movement on this? I would get on this as soon as its through review again so we could have this out by .net8!

**Comment by huoyaoyuan** [comment]
Some feedback to the API shape based on my usage:

Getting `Range` is very useful, since ROS can't be stored in collections. If I want to pass multiple segments to a function, I can pass the original ROS together with `Range[]` or `Span<Range>`.

Accessing all remaining span after enumerated for certain times is also useful.

**Comment by egbertn** [comment]
Just in case if this has not been coded yet, code below works.

`public static class MemoryExtensions
{
    public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> span, T split) where T: struct
        => new SpanSplitEnumerator<T>(span, new T[1]{ split});
    public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> split) where T: struct
        => new SpanSplitEnumerator<T>(span, split);
}

public ref struct SpanSplitEnumerator<T> where T: struct
{
    private readonly ReadOnlySpan<T> _span;
    private readonly ReadOnlySpan<T> _splitSpan;
    private readonly int maxLength;
    private readonly int _lengthOfSplit;
    private readonly int _foundSplitChars;
    private int _currentIndex = -1;
    private readonly int[] splitPositions;
    public SpanSplitEnumerator(ReadOnlySpan<T> span, ReadOnlySpan<T> split)
    {
        _span = span;
        maxLength = span.Length;
        _splitSpan = split;
        _lengthOfSplit = _splitSpan.Length;

        int currentPos = 0;
        for (; currentPos < maxLength; currentPos++ )
        {
            if (_span.Slice(currentPos, _lengthOfSplit).SequenceEqual(_splitSpan))
            {
                _foundSplitChars++;
                currentPos += _lengthOfSplit;
            }
        }
        splitPositions = new int[_foundSplitChars+1];
        //fill array
        splitPositions[0] = -1;
        _foundSplitChars =
        currentPos = 0;
        for (; currentPos < maxLength; currentPos++ )
        {
            if (_span.Slice(currentPos, _lengthOfSplit).SequenceEqual(_splitSpan))
            {
                splitPositions[++_foundSplitChars] = currentPos;
                currentPos += _lengthOfSplit;
            }
        }
    }

    public bool MoveNext() => _currentIndex++ < _foundSplitChars;

    /// <summary>
    /// the number of split sequences found
    /// </summary>
    public int Count => _foundSplitChars;

    public ReadOnlySpan<T> Current
    {
        get
        {
            int start;
            if (_currentIndex >= _foundSplitChars )
            {
                start = splitPositions[_currentIndex] + _lengthOfSplit;
                return _span.Slice(start, maxLength - start);
            }

            return _span.Slice
            (
               splitPositions[_currentIndex] + _lengthOfSplit,
                splitPositions[_currentIndex+1] - splitPositions[_currentIndex] - _lengthOfSplit
            );
        }
    }
}`

**Comment by ReubenBond** [comment]
Here's a non-allocating impl of part of API since I wanted it today:

```csharp
internal static class SpanExtensions
{
    /// <summary>
    /// Splits the span by the given sentinel, removing empty segments.
    /// </summary>
    /// <param name="span">The span to split</param>
    /// <param name="sentinel">The sentinel to split the span on.</param>
    /// <returns>An enumerator over the span segments.</returns>
    public static StringSplitEnumerator Split(this ReadOnlySpan<char> span, ReadOnlySpan<char> sentinel) => new(span, sentinel);

    internal ref struct StringSplitEnumerator
    {
        private readonly ReadOnlySpan<char> _sentinel;
        private ReadOnlySpan<char> _span;

        public StringSplitEnumerator(ReadOnlySpan<char> span, ReadOnlySpan<char> sentinel)
        {
            _span = span;
            _sentinel = sentinel;
        }

        public bool MoveNext()
        {
            if (_span.Length == 0)
            {
                return false;
            }

            var index = _span.IndexOf(_sentinel, StringComparison.Ordinal);
            if (index < 0)
            {
                Current = _span;
                _span = default;
            }
            else
            {
                Current = _span[..index];
                _span = _span[(index + 1)..];
            }

            if (Current.Length == 0)
            {
                return MoveNext();
            }

            return true;
        }

        public ReadOnlySpan<char> Current { readonly get; private set; }

        public readonly StringSplitEnumerator GetEnumerator() => this;
    }
}
```

**Comment by Phylum123** [comment]
> Here's a non-allocating impl of part of API since I wanted it today:
> 
> ```cs
> internal static class SpanExtensions
> {
>     /// <summary>
>     /// Splits the span by the given sentinel, removing empty segments.
>     /// </summary>
>     /// <param name="span">The span to split</param>
>     /// <param name="sentinel">The sentinel to split the span on.</param>
>     /// <returns>An enumerator over the span segments.</returns>
>     public static StringSplitEnumerator Split(this ReadOnlySpan<char> span, ReadOnlySpan<char> sentinel) => new(span, sentinel);
> 
>     internal ref struct StringSplitEnumerator
>     {
>         private readonly ReadOnlySpan<char> _sentinel;
>         private ReadOnlySpan<char> _span;
> 
>         public StringSplitEnumerator(ReadOnlySpan<char> span, ReadOnlySpan<char> sentinel)
>         {
>             _span = span;
>             _sentinel = sentinel;
>         }
> 
>         public bool MoveNext()
>         {
>             if (_span.Length == 0)
>             {
>                 return false;
>             }
> 
>             var index = _span.IndexOf(_sentinel, StringComparison.Ordinal);
>             if (index < 0)
>             {
>                 Current = _span;
>                 _span = default;
>             }
>             else
>             {
>                 Current = _span[..index];
>                 _span = _span[(index + 1)..];
>             }
> 
>             if (Current.Length == 0)
>             {
>                 return MoveNext();
>             }
> 
>             return true;
>         }
> 
>         public ReadOnlySpan<char> Current { readonly get; private set; }
> 
>         public readonly StringSplitEnumerator GetEnumerator() => this;
>     }
> }
> ```

There is a bug. In the case where the sentinel/delimiter is the last char, this will produce an index out of range exception.
The line " _span = _span[(index + 1)..];" should be "_span = index + 1 >= _span.Length ? default : _span[(index + 1)..];"

**Comment by khellang** [comment]
Now that we have `SearchValues<T>`, could we base this new API on top of that?

**Comment by bbartels** [comment]
> @stephentoub @GrabYourPitchforks Any chance we could get some movement on this? I would get on this as soon as its through review again so we could have this out by .net8!

Any chance we could give this another push? @stephentoub @GrabYourPitchforks ?

**Comment by stephentoub** [comment]
@jeffhandley, this has been languishing and I'd like for us to ship a solution in .NET 9. Can you help?

**Comment by GrabYourPitchforks** [comment]
Adding some comments here that I mentioned to folks offline.

If we want to use the same enumerator type as a return value from all the different overloads, then we're essentially inventing a polymorphic system. (The data storage and lookup algorithm will depend on what overload is called, and we're abstracting all of those away behind a single projection.) Polymorphism in .NET is typically implemented using a base class and derived types, but since people in this thread have said that non-allocation is vital, we can't use the typical .NET mechanisms here where state is captured by derived types which contain the concrete implementations.

So there are a few options:

- List all possible implementations' backing fields, and promote each of them to a top-level backing field of the sole `Enumerator<T>` object, and have the `MoveNext()` method switch on what algorithm is in use. This potentially bloats the size of the struct itself, and I don't know how the JIT will optimize it.
- Use a clever union-style backing field for implementation-specific details, and use a delegate or fnptr or something to jump to the appropriate implementation. This keeps the size of the struct small but adds a level of indirection, which could also impact the JIT.
- Have different method overloads return different concrete types. For example, the method which splits on a `ROS<char>` needle would return a different type than the method which splits on a `ROS<string>` needle, etc. This could limit our ability to perform certain optimizations. For example, if the API is `Split(this ROS<char> input, params ROS<char> needle)`, then `mySpan.Split()` \[which is intended to split on whitespace\] would necessarily return the same implementation as `mySpan.Split('a', 'b', 'c')`. and we might not be able to optimize it as fully as if we could return a whitespace-only specific enumerator.
- Drop the non-allocating requirement and allow internal state capture.

There are pros and cons to all of these. My previous holdup was based on a concern that whatever enumerator shape we choose for v1 of this API will have consequences for overloads and capabilities we may wish to add in the future. That is, if we decide to ship a "minimal" v1 API and we ship whatever API shape happens to fall out of that, that may unintentionally bind our ability to improve this area in v2. So I want to be sure that when the v1 API is added, the choice of shape is deliberate and is done with an eye toward future proofing. As long as that is done, I'm happy. :)

**Comment by stephentoub** [comment]
The most common use case I've seen here is to want to take an existing use of string.Split and be able to iterate through the results in a non-allocating manner. I think we should expose the APIs to do that in .NET 9, following the same structure as the Split span-based APIs we added in .NET 8, and then also have a couple to handle the generic case (in particular in support of `ReadOnlySpan<byte>` for UTF8 text):
```C#
public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> source, char separator, StringSplitOptions options = StringSplitOptions.None);
public static SpanSplitEnumerator<char> Split(this ReadOnlySpan<char> source, ReadOnlySpan<char> separator, StringSplitOptions options = StringSplitOptions.None);

public static SpanSplitEnumerator<char> SplitAny(this ReadOnlySpan<char> source, ReadOnlySpan<char> separators, StringSplitOptions options = StringSplitOptions.None);
public static SpanSplitEnumerator<char> SplitAny(this ReadOnlySpan<char> source, ReadOnlySpan<string> separators, StringSplitOptions options = StringSplitOptions.None);

public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, T separator);
public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, ReadOnlySpan<T> separator);

public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, params ReadOnlySpan<T> separators);

public ref struct StringSplitEnumerator<T>
{
    StringSplitEnumerator<T> GetEnumerator();
    bool MoveNext();
    Range Current { get; }
}
```
To fully optimize that to the max we would likely need seven different return types, and I don't believe that's warranted. The size of the returned enumerator isn't particularly important, as it's generally going to be at most one copy and it won't be put on the heap because it's a ref struct. The relevant overheads then will likely be for extra branches on each MoveNext to determine which path to take, and those should be few. Given all the work associated with split, I'm not particularly concerned about those costs. So I think we should keep this API simple and just have the single return type for all the overloads.

We've been punting on this for years. We should just do it now.

**Comment by bbartels** [comment]
Any chance I could pick this up for implementation once reviewed? I have a lot of free time to crack this out pretty quickly, especially given I'd just base in what was already (temporarily) merged. Spent a lot of time on this over the years :')

**Comment by stephentoub** [comment]
> Any chance I could pick this up for implementation once the API is reviewed? I have a lot of free time to crack this out pretty quickly, especially given I'd just base in what was already (temporarily) merged. Spent a lot of time on this over the years :')

Yes, thanks :)

**Comment by colejohnson66** [comment]
Is there a reason this isn’t called “EnumerateSplits”? Having “Split” return an enumerator, while it matches what many do with the result, doesn’t match the existing API of returning a complete result.

**Comment by stephentoub** [comment]
I'd be fine calling it EnumerateSplits, matching what we exposed on Regex.

**Comment by iSazonov** [comment]
EnumerateAllSplits, EnumerateAnySplits... Although it looks too long.

**Comment by wstaelens** [comment]
EnumerateSplitResult(s) ?

**Comment by bartonjs** [comment]
* We decided Split is better than EnumerateSplits
* We're leaving SplitAny(ROS-char, ROS-string) out for now, and only adding the generic ones.
* We added a SearchValues overload to SplitAny, though it might or might not get implemented this release
* These may need an IEquatable constraint... if they do, add it.

```C#
public static class MemoryExtensions
{
    public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, T separator);
    public static SpanSplitEnumerator<T> Split<T>(this ReadOnlySpan<T> source, ReadOnlySpan<T> separator);
    public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, params ReadOnlySpan<T> separators);
    public static SpanSplitEnumerator<T> SplitAny<T>(this ReadOnlySpan<T> source, SearchValues<T> separators);
    
    public ref struct SpanSplitEnumerator<T>
    {
        public SpanSplitEnumerator<T> GetEnumerator();
        public bool MoveNext();
        public Range Current { get; }
    }
}
```

**Comment by bbartels** [comment]
I'll get on it 🙂

**Comment by hez2010** [comment]
Should `SpanSplitEnumerator<T>` implement `IEnumerator<T>`?

**Comment by hez2010** [comment]
> It can't because it's a `ref struct`, and `ref struct`s being able to implement interfaces was removed from C# 13.

It's not being removed. It will be shipped as a preview feature instead, and several APIs in BCL have already adopted this feature.

**Comment by stephentoub** [comment]
> It will be shipped as a preview feature instead, and several APIs in BCL have already adopted this feature.

We have not implemented interfaces on any public ref structs, and we won't on non-experimental public ref structs as long as the language feature is in preview. We have no way to mark just the interface inheritance as experimental / preview, which means if the language feature were to change or disappear, we could be left with non-preview surface area we're unable to maintain.

**Comment by clipperhouse** [comment]
+1 on this overall proposal, I’m a fan of making splits into enumerators, and (hopefully) allocation-free.

I’ve implemented similar here: https://github.com/clipperhouse/uax29.net

**Comment by clipperhouse** [comment]
Having watched the design review above, on `Current` being a `Range`: another possibility is to hang a `Ranges` property off the `SpanSplitEnumerator`, as I’ve done [here](https://github.com/clipperhouse/uax29.net/blob/main/uax29/Tokenizer.cs#L139). That way, you can preserve the API of `string.Split`, where the enumerator returns the splits, but give consumers the ranges to use where that’s advantageous.

Though I am not sure that satisfies the requirements mentioned in the discussion. IIUC, the goal was to a) allow `Memory` inputs to avoid allocation and b) perhaps it’s confusing if the input is `Memory` but the output is `ReadOnlySpan` or c) avoid multiplying the overloads.

(I appreciate that you likely feel the API has been sufficiently debated and thanks for indulging.)

**Comment by stephentoub** [comment]
Fixed by https://github.com/dotnet/runtime/pull/104534
